#69 Add architecture document. (#75)
tcdi/plrust: doc/architecture.md

# Architecture

This document describes the high-level architecture of plrust.
If you want to familiarize yourself with the code base, you are just in the right place!

## Bird's Eye View

![](assets/architecture_1.png)

> 

Reviewer: Maybe this is neither here nor there, but I am wondering how far we could get with the new [mermaid](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/) syntax instead of image files? 

On one hand, it'd be cool to not have binary file revisions in the git history every time the architecture needed to change, but on the other hand it'd be another syntax to learn. I am torn.

## Design Goals

Design a custom rust compilation target for Postgres that provides nearly "safe" (as Rust defines it) and "trusted" (as Postgres defines a procedural language) plrust.

The goals for the approach include

* Architecture support for x86_64 and aarch64
* Operating system support for Linux
* Disallow File Handle operations
* Disallow access to the internals of the database
* Disallow access to the OS as the user executing the Postgres process 
* Disallow access into active postmaster process, i.e. no ability to reach into Postgres memory space, despite executing inside it.
* Gracefully handle rust panics and have them interoperating with Postgres' transaction system

> * Memory allocation within Postgres' pallow/pfree functions

Reviewer 2: s/pallow/palloc/

# Architecture

> `plrust`, a rust based [Postgres Extension for Postgres](https://github.com/tcdi/pgx), provides a rust language handler for Postgres Functions. The installed language handler compiles postgres function, written in rust, executes the function <where>. 

Reviewer 2: ````suggestion
`plrust`, a Rust-based [extension for Postgres](https://github.com/tcdi/pgx), provides a procedural language handler for Rust in Postgres. When installed, `plrust` allows this to work:
```sql
CREATE FUNCTION {fn_name} ({args})
RETURNS {ret}
-- function attributes can go here
AS $$
    // PL/Rust function body goes here
$$ LANGUAGE plrust;
```

PL/Rust will compile the function as a dynamic library, load it, and execute it.
````

## Trusted Language

> In order to create a [trusted](https://www.postgresql.org/docs/current/sql-createlanguage.html) language handler in Postgres we must restrict the language handler from specific operations such as 

Reviewer 2: ```suggestion
In order to create a [trusted](https://www.postgresql.org/docs/current/sql-createlanguage.html) language handler in Postgres we must restrict the functions compiled and executed by the language handler to the set of operations other code in Postgres has access to.
```

> - File Handle Operations

Reviewer 2: ```suggestion
- No operations on Files except through the database itself
```

- Limit access to the database to that of other procedural language functions
- Limit access to system resources to those of a trusted language user function

> - Any unprivileged database user can be permitted to use the language ([postgresql.org](https://www.postgresql.org/docs/current/plperl-trusted.html))

Reviewer 2: ```suggestion
- It must be sound to allow any unprivileged database user to use the language ([postgresql.org](https://www.postgresql.org/docs/current/plperl-trusted.html))
```

* Disallow access to the OS as the user executing the Postgres process 
* Disallow access into active postmaster process, i.e. no ability to reach into Postgres memory space, despite executing inside it.
* Gracefully handle rust panics and have them interoperate with Postgres' transaction system
* Memory allocation within Postgres' palloc/pfree functions

## Approach

The plrust extension is compiled using the standard rust library.  The postgrestd library is used to compile the functions written using PLRust.  The postgrestd library is a libraries are compiled using The 

Following an approach similar to the selection between libc and the musl libc standard library for compilation, a Postgres compilation target is defined that instructs the compiler to use the postgrestd library.  The postgrestd library provides the rust standard library interfaces except in the case where it is desirable to prevent access.  In those cases the code is [configured](https://doc.rust-lang.org/stable/rust-by-example/attribute/cfg.html) to be not present. The result is a small shim on top of the rust library limited access to the standard library.

## Bird's Eye View

![](assets/architecture_1.png)

> ## Code Map
> 
> This section talks briefly about various important directories and data structures.

Reviewer 2: ```suggestion
## Supporting Crates

Because PL/Rust implements a fairly complicated language and makes it sound to use as a trusted procedural language, there are multiple supporting crates necessary to make it work.

### pgx with `features = ["postgrestd", ..]`

The `plrust` language handler is itself implemented using pgx, but also all PL/Rust functions must understand the PostgreSQL data types in the same way that `pgx` enables. In addition, bindings via `pgx` are used to implement most of the other supporting crates.
```

### pallocator

The Postgres allocator project maps the Postgres memory allocation methods to standard library methods.  the memory allocation from standard library methods to postgres specific methods.

alloc -> palloc
free -> pfree
realloc - prealloc

### postpanic

The Postgres panic project maps the rust panic to the postgres panic, allowing Postgres to handle the panic within the transaction system.

> ### std

Reviewer 2: ```suggestion
### postgrestd

See [postgrestd](https://github.com/tcdi/postgrestd) for more details.
```

- cannot gain OS level access with the permissions of the server process.
- Any unprivileged database user can be permitted to use the language ([postgresql.org](https://www.postgresql.org/docs/current/plperl-trusted.html))

## Rust

> While interpreted languages like Perl or Python execute in a runtime component, the Rust language must be compiled to a specific target.  A target can be a shared object, binary, or wasm artifact.  The operating system or host process executes the shard object (or binary) in a similar manner to C.  Rust provides support for unsafe memory operations through the use of the unsafe keyword.  Code tagged in this manner indicates to the compiler that the code does not need to be checked for memory safety and the developer has ensured it safety.

Reviewer 2: The extension called "PL/Rust" which includes the language handler is responsible for covering the linking and loading steps. This extension may have privileges that user functions do not, using the Rust std of the same host target that PostgreSQL itself is compiled for, to interoperate in that privileged mode.
